### ISSUE TITLE
GSNv2 support

### ISSUE THREAD

**Issue by spalladino** [OpenZeppelin/openzeppelin-contracts#2258]
**🧐 Motivation**
[GSNv2](https://docs.opengsn.org/learn/index.html) is under development by the OpenGSN team, and it's very similar to GSNv1 regarding recipient contract implementation. We could add support for both v2 and v1 with some minor changes.

**📝 Details**
GSNv2 moves the responsibility of deciding whether to pay for a meta-tx (ie `acceptRelayedCall`, along with `pre/post`) out of the Recipient into a [Paymaster contract](https://docs.opengsn.org/contracts/index.html#paymaster). The Recipient now only needs to implement the base `_msgSender` and `_msgData` helpers, which should use the injected address at the end of the CALLDATA (instead of msg.sender) if the call [comes from a **TrustedForwarder**](https://docs.opengsn.org/contracts/index.html#trusted_forwarder).

The TrustedForwarder has the responsibilities of validating the user signature, managing its nonce, appending the signer address to the CALLDATA, and forwarding the call. As with the RelayHub in GSNv1, the Recipient should assume they are receiving a meta-tx if it's sent by the TrustedForwarder. And as with the RelayHub, the TrustedForwarder can be a singleton, since it has no app-specific logic.

This approach is described in detail in [this ERC to be published](https://docs.google.com/document/d/1EnLof0KCtdnrqp07MN6m3TQWDKVrjYMKttL9yNW2pY4/edit?usp=sharing). Note that `getHubAddr` was changed to [`isTrustedForwarder`](https://github.com/opengsn/gsn/blob/master/contracts/interfaces/IRelayRecipient.sol#L15) in the Recipient internal interface.

To sum up, we need to:
- Refactor the _msgSender and _msgData methods out of [GSNRecipient](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/GSN/GSNRecipient.sol) into a MetaTransactionRecipient (preferrably including the ERC number once published).
- Add a TrustedForwarder implementation to Contracts so devs can easily set up a local meta-tx setup, and optionally include addresses to the singleton deployments.
- Clarify that GSNRecipient is only valid for GSNv1, since the [Paymaster interface](https://github.com/opengsn/gsn/blob/master/contracts/interfaces/IPaymaster.sol) has changed in v2.

**Comment by nventuro** [comment]
Great, thanks for the summary! It looks like we'll be able to leave our `Context` contract untouched, and therefore not modify any non-GSN code. The 'issue' introduced by diamond inheritance that requires doing a manual override of ``Context``'s methods will remain though.

We may also want to update our docs to guide users to the OpenGSN docs, removing the references we have (which are anyway replicated there).

**Comment by naszam** [comment]
Would be great if OpenZeppelin Test Environment could support testing OpenGSNv2, cause currently the only way to test Paymaster and Meta Txs can be done via Truffle Tests or Builder following GSNv2 Docs.
I've posted a related issue on OpenZeppelin Forum [here ](https://forum.openzeppelin.com/t/automated-tests-for-gsnv2-meta-txs-via-openzeppelin-test-environment/3898).
Thanks!

**Comment by frangio** [comment]
@naszam For testing please open an issue in [OpenZeppelin/openzeppelin-test-environment](https://github.com/OpenZeppelin/openzeppelin-test-environment).

**Comment by naszam** [comment]
@frangio Ok, Thanks! I've removed my post
